compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUSelectUKernels.cpp

```diff
@@ -158,39 +170,178 @@ static ArrayAttr lookUpExecutableObjects(Operation *op,
+
+  auto bitcodeData = bitcodeObject.getData();
+  std::string buffer;
+  buffer.resize(bitcodeData.getStorageSize());
+  if (failed(bitcodeObject.getData().serializeToBuffer(
+          op->getLoc(), llvm::endianness::native,
+          ArrayRef<char>{buffer.data(), buffer.size()}))) {
+    op.emitWarning("Failed to serialize bitcode.");
+    return {};
+  }
+  llvm::LLVMContext llvmContext;
+  llvm::Expected<std::unique_ptr<llvm::Module>> module =
+      llvm::getLazyBitcodeModule(
+          llvm::MemoryBufferRef{buffer, ukernelConfig.getName()}, llvmContext,
+          /*ShouldLazyLoadMetadata=*/true);
+  if (!module) {
+    op.emitWarning("Failed to parse bitcode module.");
+    return {};
+  }
+  llvm::EngineBuilder builder(std::move(module.get()));
+  std::string builderError;
+  builder.setEngineKind(llvm::EngineKind::Interpreter)
+      .setErrorStr(&builderError);
+  std::unique_ptr<llvm::ExecutionEngine> interpreter{builder.create()};
+  if (!interpreter) {
+    op.emitWarning("Failed to create the interpreter.");
+    return {};
+  }
+  llvm::Function *func = interpreter->FindFunctionNamed(
+      llvm::formatv("{}_query_shared_memory_size", ukernelConfig.getName())
+          .str());
+  auto constI32 = [](int32_t val) {
+    llvm::GenericValue v;
+    v.IntVal = APInt(32, val);
+    return v;
+  };
+  llvm::GenericValue c = interpreter->runFunction(
+      func, {constI32(mma.getIntrinsicsM()), constI32(mma.getSubgroupsM()),
+             constI32(mma.getIntrinsicsN()), constI32(mma.getSubgroupsN()),
+             constI32(mma.getIntrinsicsK())});
+  int sharedMemorySize = c.IntVal.getSExtValue();
+
+  // Reject a ukernel that would consume too much shared memory, which we need
+  // to save for other purposes. This threshold can always be adjusted but we
+  // default to a low threshold to get an early signal.
+  int maxSharedMemorySize = getSharedMemorySize(gpuTarget) / 4;
+  if (sharedMemorySize > maxSharedMemorySize) {
+    op.emitWarning(llvm::formatv("The shared memory size {} required by the "
+                                 "ukernel exceeds the maximum allowed size {}.",
+                                 sharedMemorySize, maxSharedMemorySize));
+    return {};
+  }
+  return sharedMemorySize;
+}
+
+// Returns the shared memory size required by the multi_mma ukernel.
+// On failure, an op warning is emitted and {} is returned.
+// Uses a static cache to avoid calling expensivelyEvaluateSharedMemorySize
+// more than once per DataTiledMMAAttr value.
+static std::optional<int>
+getSharedMemorySize(IREE::GPU::MultiMmaOp op,
+                    IREE::GPU::UKernelConfigAttr ukernelConfig,
+                    IREE::HAL::ExecutableObjectAttr bitcodeObject,
+                    IREE::GPU::TargetAttr gpuTarget) {
+  auto mma = dyn_cast<IREE::GPU::DataTiledMMAAttr>(op.getKind());
+
+  // The cache and the mutex guarding it.
+  static llvm::DenseMap<IREE::GPU::DataTiledMMAAttr, std::optional<int>>
+      cachedSharedMemorySizes;
```

Comment:
Can we cache it like this without any cleanup on context exit? I'm thinking about a scenario where iree compiler is used as a library with multiple context objects, some of which may be destroyed (which will result in invalidated pointers).

Reply:
Can you elaborate? How is the context object entering the picture?

Reply:
Say we have a python program that invokes IREE. We perform the first 100 compilation with MLIRContext number 1 and then 100 compilation s with MLIRContext number 2. When we compile for the second context, the previous one might have already been destroyed and relying on pointer values from it is UB.

A scenario where this may be reasonable is a tuner or a some compilation server that needs to periodically reset the context to prevent it form growing in size indefinitely.

Reply:
The alternative would be to have it stored in a dialect, e.g.: https://github.com/iree-org/iree/blob/3f713f5f4743ddb6715e4ea4a361784b54489e5a/compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.td#L39-L68

Reply:
Clever... but way easier still to just stringify the keys!